lib/hanami/application.rb

```diff
@@ -195,13 +195,23 @@ def prepare_container_base_config
         ]
       end
 
-      def prepare_container_component_dirs
-        return unless root.join(APP_DIR)&.directory?
+      def prepare_container_component_dirs # rubocop:disable Metrics/AbcSize
+        return unless root&.directory?
 
-        # Expexct component files in the root of the app/ component dir to define classes
-        # inside the application's namespace
-        container.config.component_dirs.add(APP_DIR) do |dir|
-          dir.namespaces.add_root(key: nil, const: application_name.name)
+        # Component files in both `app/` and `app/lib/` define classes in the
+        # application's namespace
+
+        if root.join(APP_DIR, LIB_DIR).directory?
+          container.config.component_dirs.add(File.join(APP_DIR, LIB_DIR)) do |dir|
```

Comment:
@timriley any specific reason why we use `root.join` for the `if` statement and then `File.join` for adding the dir to the config?

Is that because we need the absolute dir to check its existence and then the relative dir to pass to dry-system?

Reply:
@jodosha 100% correct 😄